library/std/src/ffi/c_str.rs

```diff
@@ -1266,34 +1266,40 @@ impl CStr {
     ///
     /// ```no_run
     /// # #![allow(unused_must_use)] #![allow(temporary_cstring_as_ptr)]
-    /// use std::ffi::CString;
+    /// use std::ffi::CStr;
     ///
-    /// let ptr = CString::new("Hello").expect("CString::new failed").as_ptr();
+    /// let ptr = {
+    ///     let bytes = *b"Hello\0";
```

Comment:
I think this example is a bit too subtle, since without the `*` here (which I missed the first time I read this) nothing will dangle here.

Maybe it's fine to keep the original example with `CString`, since `.as_ptr()` on a `CString` just calls `CStr::as_ptr()`, and it's probably the clearest example we have of making a `&CStr` to heap allocated data.

Reply:
The original example didn't actually use the function that is being documented, however—which is the issue this PR was trying to address :-)  How about we extract a `&CStr` from `CString::as_c_str` and then call `as_ptr` on that?  To be honest, it all seems pretty contrived whichever way we go... and maybe this lifetime issue is actually so obvious with a `CStr` that it doesn't even need to be explicitly documented?

Reply:
> The original example didn't actually use the function that is being documented

It did. :-)

`CString` does not have a `as_ptr()`. Calling `as_ptr()` on a `CString` just calls `CStr::as_ptr()` through `CString`'s `Deref` implementation. See https://doc.rust-lang.org/std/ffi/struct.CString.html#deref-methods-CStr.

Reply:
I've created a block in which a `CString` is declared, then a `CStr` obtained, on which `.as_ptr()` is called—so that the steps are explicitly clear.